### PR TITLE
feat: mutable route tree

### PR DESCRIPTION
**Pull request by jacob-ebey**

### CODE REVIEW

File: packages/router/router.ts [hunk 1]
```diff
@@ -1108,14 +1123,15 @@ export function createRouter(init: RouterInit): Router {
     saveScrollPosition(state.location, state.matches);
     pendingPreventScrollReset = (opts && opts.preventScrollReset) === true;
 
+    let routesToUse = inFlightDataRoutes || dataRoutes;
```
Comment (brophdawg11) [inline packages/router/router.ts:1126]: This feels sufficient for now since it's only needed in a few spots.  But I am wondering if this feels like a hard-to-detect footgun down the road if we add some new code and just reach for `dataRoutes`.  One thought was a quick getter function like `getCurrentDataRoutes()` but maybe even something like renaming `dataRoutes -> comittedRoutes` would help add some explicit semantics too.  I think it's probably best to table the decision until we tackle the follow ups for fog of war and MFE though

File: packages/router/__tests__/router-test.ts [hunk 2]
```diff
@@ -362,6 +364,9 @@ function setup({
       return enhancedRoute;
     });
   }
+  if (ref) {
+    ref.enhanceRoutes = enhanceRoutes;
```
Comment (brophdawg11) [inline packages/router/__tests__/router-test.ts:368]: If we go this `enhanceRoutes` approach I would just add it to the object returned from `setup` and avoid the need for the `ref`?

File: packages/router/__tests__/router-test.ts [hunk 3]
```diff
@@ -13544,4 +13555,60 @@ describe("a router", () => {
...
+              path: "/foo",
+              id: "foo",
+              loader: false,
+              action: true,
+            },
+          ],
+        },
+      ]);
+      t.router._internalSetRoutes(newRoutes);
+      t.router.revalidate();
```
Comment (brophdawg11) [inline packages/router/__tests__/router-test.ts:13594]: I think you can just lift the `let R = await t.revalidate();` up to here so there's only one revalidation happening?

File: packages/router/__tests__/router-test.ts [hunk 4]
```diff
@@ -13544,4 +13555,60 @@ describe("a router", () => {
...
+      // Get a new revalidation helper that should use the updated routes
+      let R = await t.revalidate();
+      // Should still be og roues on new revalidation as one started by update
+      // has not yet completed
+      expect(t.router.routes).toBe(ogRoutes);
+      // Resolve any loaders that should have ran
+      await R.loaders.root.resolve("ROOT*");
+      // Don't resolve "foo" because it was removed
+      // Revalidation should be complete
+      expect(t.router.state.revalidation).toBe("idle");
```
Comment (brophdawg11) [inline packages/router/__tests__/router-test.ts:13608]: I'd assert `loaderData` here too to ensure `foo` got removed completely

**Review by brophdawg11 — COMMENTED** [review]
LGTM!  Let's add some edge case unit tests?  We have the "removed a loader during HMR", but probably want "added a loader" as well as a few around revalidation being interrupted by a navigation + fetch

File: packages/router/router.ts (hunk 1 above)
Comment (pcattori) [reply packages/router/router.ts:1126]: Yea let's tackle this for sure when we got to the granular route updates API

File: packages/router/__tests__/router-test.ts (hunk 3 above)
Comment (jacob-ebey) [reply packages/router/__tests__/router-test.ts]: Thanks, was trying to figure out how to use this API.

File: packages/router/__tests__/router-test.ts (hunk 2 above)
Comment (jacob-ebey) [reply packages/router/__tests__/router-test.ts]: removed ref, return enhanceRoutes, add helper for _internalSetRoutes

**Review by pcattori — APPROVED** [review]

**Review by brophdawg11 — APPROVED** [review]
